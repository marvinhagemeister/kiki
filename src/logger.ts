import { IFile } from "./io/file";
import chalk from "chalk";
import { SassError } from "node-sass";

/* tslint:disable no-console */

export interface LoggerAdapter {
  log(...args: any[]): void;
  error(...args: any[]): void;
}

export class Logger {
  constructor(private adapter: LoggerAdapter = console) {}

  start(name: string, path: string) {
    const l = this.adapter;
    l.log();
    l.log(chalk.white(`=> compiling ${name} from: `) + chalk.yellow(path));
    l.log();
  }

  watch(path: string[]) {
    const l = this.adapter;
    l.log(chalk.white("=> watching: ") + chalk.yellow(path.join(", ")));
    l.log();
  }

  nothingToDo() {
    const l = this.adapter;
    l.log();
    l.log(chalk.blue("Nothing to to do!"));
    l.log();
  }

  time(file: string, time: string | number) {
    const l = this.adapter;
    l.log("  " + chalk.dim(file) + " " + chalk.blue(time + "ms"));
  }

  invalidJson(message: string, path: string) {
    const l = this.adapter;
    l.log(chalk.red("Invalid JSON in " + path));
    l.log("  " + chalk.red("> " + message.replace(/^\/.+\:\s/, "")));
    l.log();
  }

  noFilesOrOnlyPartials() {
    const l = this.adapter;
    l.log();
    l.log(
      chalk.yellow(
        "Only partials found. Partials must be imported by " +
          "a root file to be included in compilations.",
      ),
    );
    l.log();
  }

  error(err: SassError | Error | string) {
    const l = this.adapter;

    l.log();
    if (typeof err === "string" || err instanceof String) {
      l.error(chalk.red(err.toString()));
    } else {
      if (err.message.indexOf("search path") > -1) {
        l.error(chalk.red(err.message));
      } else if ((err as any).formatted !== undefined) {
        l.error(chalk.red((err as any).formatted));
      } else {
        l.error(chalk.red(err.stack || ""));
      }
    }

    l.log();
  }

  warning(warn: any) {
    const l = this.adapter;
    l.log(chalk.red(warn.text));
    l.log(chalk.white(warn.node));
    l.log();
  }

  success(msg: string) {
    const l = this.adapter;
    l.log("  " + chalk.green(msg));
  }

  taskDone(files: IFile[], time: number) {
    const l = this.adapter;
    l.log();
    l.log("  " + chalk.white("time: ") + chalk.blue(time + "ms"));
    l.log();

    files.forEach((file, i) => {
      const name = file.location.replace(process.cwd() + "/", "");
      l.log("  " + chalk.dim(name));
    });

    l.log();
  }

  change(event: string, path: string) {
    const l = this.adapter;
    event = event === "change" ? "modified" : event;
    l.log(event + ": " + path);
  }
}

const logger = new Logger(console);
export default logger;
