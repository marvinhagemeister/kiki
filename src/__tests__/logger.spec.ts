import { Logger } from "../logger";
import { StubLogger } from "./helpers";

describe("Logger", () => {
  const stub = new StubLogger();
  const l = new Logger(stub);

  afterEach(() => (stub._output = ""));

  it("should start", () => {
    l.start("foo", "/my/path");
    expect(stub.output).toMatchSnapshot();
  });

  it("should watch", () => {
    l.watch(["/foo", "/my/path"]);
    expect(stub.output).toMatchSnapshot();
  });

  it("should nothing to do", () => {
    l.nothingToDo();
    expect(stub.output).toMatchSnapshot();
  });

  it("should time", () => {
    l.time("/foo", 123);
    expect(stub.output).toMatchSnapshot();
  });

  it("should invalidJson", () => {
    l.invalidJson("nope", "/foo");
    expect(stub.output).toMatchSnapshot();
  });

  it("should noFilesOrOnlyPartials", () => {
    l.noFilesOrOnlyPartials();
    expect(stub.output).toMatchSnapshot();
  });

  it("should warning", () => {
    l.warning("foo");
    expect(stub.output).toMatchSnapshot();
  });

  it("should success", () => {
    l.success("foo");
    expect(stub.output).toMatchSnapshot();
  });

  it("should taskDone", () => {
    l.taskDone(
      [{ base: "foo", content: "content", location: "log", map: "map" }],
      123,
    );
    expect(stub.output).toMatchSnapshot();
  });

  it("should change", () => {
    l.change("foo", "/foo");
    expect(stub.output).toMatchSnapshot();
  });
});
