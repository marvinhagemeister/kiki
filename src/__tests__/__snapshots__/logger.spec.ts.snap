// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should change 1`] = `
"foo: /foo
"
`;

exports[`Logger should invalidJson 1`] = `
"Invalid JSON in /foo
  > nope

"
`;

exports[`Logger should noFilesOrOnlyPartials 1`] = `
"
Only partials found. Partials must be imported by a root file to be included in compilations.

"
`;

exports[`Logger should nothing to do 1`] = `
"
Nothing to to do!

"
`;

exports[`Logger should start 1`] = `
"
=> compiling foo from: /my/path

"
`;

exports[`Logger should success 1`] = `
"  foo
"
`;

exports[`Logger should taskDone 1`] = `
"
  time: 123ms

  log

"
`;

exports[`Logger should time 1`] = `
"  /foo 123ms
"
`;

exports[`Logger should warning 1`] = `
"undefined
undefined

"
`;

exports[`Logger should watch 1`] = `
"=> watching: /foo, /my/path

"
`;
